### ISSUE TITLE
Training Language Models with Memory Augmentation

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#25]
# Training Language Models with Memory Augmentation

# 개요

- Memory Augmentation 방법론은 언어 모델의 파라미터나 계산량을 늘리지 않고 성능을 향상시킬 수 있음.
    - 최근 입력의 neural cache 형태의 메모리 유닛에 접근하거나, 외부 look-up table을 이용하는 방식
- 그러나 기존의 방법론들은 대부분 testing time에만 메모리 유닛이 도입되거나, 별도로 학습된 모델을 이용하기 때문에 학습 과정에서 직접적으로 옵티마이징되지 않아 성능이 sub-optimal할 가능성이 존재
- 본 논문에서는 새로운 memory augmentation 방법론인 TRIME(Training with In-batch Memories)를 제안
    - contrastive learning에서 영감을 받아, in-batch examples을 accessible memory로써 활용함.
- 본 논문에서는 언어 모델링에서 활용할 수 있는 memory의 종류를 3종류로 제시함.
    1. local memory: attention을 통해 모델링 가능한 최근 과거의 컨텍스트 정보
    2. long-term memory: 동일한 문서에 등장하지만 입력 길이의 제한으로 인해 직접적으로 접근할 수 없는 컨텍스트 정보
    3. external memory: 전체 학습셋에 저장되어 있거나 외부 코퍼스에 존재하는 정보
    - 위의 3가지 메모리 정보를 잘 활용하기 위한 학습시의 새로운 데이터 배칭 전략을 제시
- 모델 아키텍쳐 변경 없이 다양한 언어 모델링, 기계 번역 태스크에서 성능 향상을 확인함.

# Preliminaries

## Language Modeling

- 본 논문에서는 주로 언어 모델링에 초점을 맞추지만, 방법론 자체는 일반적인 텍스트 생성 태스크 전반에 적용 가능함.
- 뉴럴 언어 모델은 $c_t=x_1,...,x_{t-1}$의 컨텍스트를 입력받아 이를 $f_{\theta}(c_t) \in \mathbb{R}^d$의 벡터 표현으로 매핑함.
- 다음 토큰 예측 확률은 아래와 같이 표현
    
    $$
    P(w|c_t) \propto exp(E_w^Tf_{\theta}(c_t))
    $$
    
    - $E_w \in \mathbb{R}^d$은 토큰 $w \in V$의 아웃풋 임베딩을 의미
- 파라미터는 ground truth 토큰들의 negative log-likelihood를 최소화하는 방향으로 최적화

## Memory Augmentation

- 메모리를 $\{(c_i,x_i)\}$와 같은 컨텍스트-타겟의 페어 집합으로 간주

### Local Memory

- 동일 인풋내의 가장 최근 토큰들을 의미
    - $c_t=x_1,...,x_{t-1}$

$$
M_{local}(c_t)=\{(c_j,x_j)\}_{1\leq j\leq t-1}
$$

- 간단히 예를 들어서 L=”abcded”의 시퀀스가 있다고 하면
    
    $M_{local}(L)=\{(a,b),(ab,c),(abc,d),(abcd,e),(abcde,d)\}$
    
- LSTM 같은 모델은 인퍼런스시에 학습시의 시퀀스 길이보다 더 긴 시퀀스 길이를 입력하기 용이함. (그래서 “continous cache”라고 부름)
- 트랜스포머 모델로 치면 attention을 통해 접근 가능한 이전 토큰들

### Long Term Memory

- 동일 문서 내의 컨텍스트이나 attention을 통해서는 접근할 수 없는 long-range context
    - 트랜스포머 모델은 연산량의 한계 때문에 100-1k의 토큰 밖에 모델에 입력할 수 없기 때문
- 문서를 $s^{(1)},...,s^{(T)}$로 분절한다고 했을 때, 한 세그먼트 $s^{(i)} = \{c_1^{(i)},...,c_L^{(i)}\}$ 는 L 길이의 컨텍스트를 포함하고 있음.
- 이 때, long-term memory는 아래와 같이 표현 가능

![스크린샷 2022-08-06 오후 11 31 22](https://user-images.githubusercontent.com/49383323/183253131-3c480f2b-7b0e-4105-a03f-ce7b8896196c.png)

- 기존 연구들에서는 이전 세그먼트들의 hidden representations을 modified Transformer 구조에 활용하여 long-range dependency를 모델링하고자 했음.
- 본 논문의 방법론은 모델 아키텍쳐 변경 없이 적용 가능

### External Memory

- 외부 메모리는 대형 코퍼스 $\mathcal{D}$를 가정 (단순히 하면 학습 코퍼스)

$$
M_{ext}=\{(c_j,x_j) \in \mathcal{D}\}
$$

- 해당 메모리는 위의 두 메모리보다 압도적으로 거대한 양: (e.g., 10^8)
    - 해당 메모리의 모든 메모리에 접근하는 건 매우 비싸기 때문에 주로 approximate nearest neighbor search를 통해 접근함.

# Training with In-batch Memories

<img width="454" alt="1" src="https://user-images.githubusercontent.com/49383323/183253215-6ec9997b-c570-4d77-9419-ac7f6230dd34.png">

- 일반적인 언어 모델 학습과 달리, training memories $M_{train}=\{(c_i,x_i)\}$를 가정.
    - 학습하는 동안 동일 학습 배치내에서 on-the-fly로 만들 수 있기 때문에 backprop이 가능함.
- TRIME에서는 다음 토큰 예측 확률을 아래와 같이 정의
    
   <img width="454" alt="2" src="https://user-images.githubusercontent.com/49383323/183253225-2310b53f-bd1f-4b9a-84e1-ca9ea210fe1b.png">

    - $f_{\theta}(c)$는 트랜스포머 모델의 output representation
    - $E_w$는 토큰 임베딩
    - $g_{\theta}$는 $c$와 메모리 내의 모든 컨텍스트 $c_j$와의 유사도를 구하는데 쓰는 representations
        - 일반적으로는 $g_{\theta}=f_{\theta}$로 둘수 있으나, 경험적으로 트랜스포머의 마지막 feed-forward layer의 인풋을 사용하는게 제일 좋았다고 함.
        - 유사도는 $sim(q,k)=\frac{qk}{\sqrt{d}}$와 같은 scaled dot product를 이용
- 위와 같은 training objectives는 contrastive loss의 프레임워크로 해석할 수 있음.
    - query representation $f_{\theta}(c)$와 *static* token representation $E_{w*}$를 얼라인시키는 것과
    - query representation $f_{\theta}(c)$와 동일한 다음 토큰을 갖는 *contextualized* representation (i.e. $g_{\theta}(c_j)$ for $x_j=w^*$)을 얼라인시키는 것
        - 즉, 메모리 내에서 동일한 토큰을 예측하는 컨텍스트들을 positives, 나머지를 negatives로 하는 contrastive loss 항을 추가한 것.
        - (다운) 다시 위의 예를 가져와서, L=”abcded”의 시퀀스가 있다고 하고 $P(d|abcde)$의 확률을 구한다고 하면, 아래의 메모리 중에서 $(abc,d)$의 예시가 positive가 되고 나머지는 negatives가 된다.
            
            $M_{local}(L)=\{(a,b),(ab,c),(abc,d),(abcd,e),(abcde,d)\}$
            
    - 만약 $w^*$가 레어 워드가 training memory에 없다면 objective는 그냥 단순 언어모델링 태스크가 됨
- 이렇게 되면, 모델에 입력되는 컨텍스트 밖의 컨텍스트도 학습에 직접적으로 관여할 수 있게 된다.

# Adaption to Different Memories

### Inference

- 위에서 설명한 3가지 타입의 메모리를 테스트시에 어떻게 이용할 수 있을까?
- testing objective도 training objective와 기본적으로 동일
    - 대신 testing memories를 $M_{local}, M_{long},M_{ext}$의 조합으로 이용함.

### Notation

- $L$: 세그먼트 길이
- $B$: 훈련 배치에 사용된 세그먼트의 개수
- $m$: 배치 내 각 문서의 연속적인 세그먼트의 개수
    - 각 배치는 $b \approx \frac{B}{m}$개의 서로 다른 문서를 포함함.
- 위의 3가지 하이퍼파라미터들은 학습시에는 최적의 파라미터를 설정하고, 인퍼런스시에는 자유롭게 변경해볼 수 있음.
- key challenge는 테스팅 메모리가 매우 커질 수 있다는 것
    - $M_{long}$은 10^4 정도, $M_{exp}$는 10^8 정도로 클 수 있음.
    - 학습 메모리를 테스팅 메모리와 동등한 수준으로 유지할 수가 없음.
- 이후 섹션에선, 학습과 테스팅 간의 discrepancy를 줄이기 위한 training memories 구축 방법 및 데이터 배칭 방법을 서술함.
    - TRIME-LM
    - TRIME-LM_long
    - TRIME-LM_ext

<img width="1143" alt="3" src="https://user-images.githubusercontent.com/49383323/183253238-79801cf2-65b5-4e59-b9d7-62f18249fb5d.png">

## Local Memory

- 동일 세그먼트 내의 이전 토큰들
- 일반적인 언어모델링 학습과 동일
- training objective를 아래와 같이 사용하는 것만 다름
    
<img width="454" alt="4" src="https://user-images.githubusercontent.com/49383323/183253295-2a5fb758-0178-4ec7-a675-aa73b54f8007.png">

- $x_1,...,x_L$을 인코딩 하는 연산량에 비하면 추가적인 computational overhead는 매우 적은 편
- 이 방식을 TRIMELM이라 부름.
- 이 방식만으로도 LM 벤치마크 태스크들에서 notable gain을 보여줌

## Long-Term Memory

- long-term memory augmentation을 위해서 훈련 배치 내에 동일 문서의 연속적인 segment를 packing 함.
- Figure 2(b)와 같은 방식
- 테스트할때는 학습때보다 더 긴 컨텍스트도 사용할 수 있는데, dev set을 이용하여 optimum 값을 결정했다고 함.
- 이 방식을 TRIMELM_long이라 부름.
- 기존 연구들은 이전 segments의 hidden representations를 저장하거나 모델 구조를 수정해야 하는 등 현실적으로 적용하기 어려운 문제들이 있었으나, 본 방법론은 모델 구조를 수정할 필요 없고, last layer의 아웃풋만 이용하므로 적용하기 쉬움.
- 또한 기존 연구들에서는 stale memory representations (최신 모델이 아닌 이전 스텝의 representations)을 이용해야하거나, 이전 세그먼트에 back-prop할 수 없다는 문제가 있었으나, 본 방법론은 back-prop이 가능하고, 동일 모델의 memory representations을 사용

## External Memory

- 외부 메모리는 전체 훈련셋과 같은 큰 코퍼스의 context-target pairs를 포함하고 있으므로 similarity search를 통해 top-K 페어를 리트리벌하여 사용
- 학습용 배치를 만들 때, BM25를 통해 lexical overlap이 큰 세그먼트들을 동일 배치내에 패킹함.
    - 처음에 single segment에서 시작해서, 반복적으로 highest BM25 스코어를 갖는 세그먼트들을 동일 배치내에 포함시킴.
- 현재 세그먼트와 여타 세그먼트를 학습에 동시에 이용하게 되면, 모델이 local memory만을 활용하는 경향이 생김.
    - 따라서, 학습할 때 p=90%의 확률로 로컬 메모리를 exclude하여 기타 세그먼트를 참조할 수 있게 encourage함.
    - 이 방식이 significantly 성능 향상을 가져옴.
    - (다운) 우리도 무작정 시퀀스 길이를 늘리는게 능사가 아닐수도 있다는 느낌도 든다. 시퀀스 길이가 길어지는 효과를 충분히 보려면 앞의 시퀀스를 참조해서 추론하는 방법을 학습시켜야 성능 향상이 눈에 띄게 있지 않을까 싶음.
        - 이 논문에선 변형된 objectives와 위와 같은 트릭을 이용해서 앞의 시퀀스를 참조하는 방법을 가르쳤다.

# 실험

## 데이터셋 및 태스크

- Wikitext-103
    - 103M 토큰으로 구성된 word level 언어 모델링 데이터셋
    - 247M 트랜스포머 모델 segment length L=3,072
    - 150M 트랜스포머 모델 segment length L= 150
- Enwik8
    - 100M 토큰으로 구성된 char level 언어 모델링 데이터셋
    - 12-layer 트랜스포머 모델 segment length L=512
- IWSLT’14
    - 170K 번역 페어로 구성된 DE→EN 기계 번역 태스크
    - Transformer encoder-decoder 모델

## 결과

<img width="1125" alt="5" src="https://user-images.githubusercontent.com/49383323/183253325-375f8859-9e64-4e74-aaba-cc1ab53a7bab.png">

<img width="575" alt="6" src="https://user-images.githubusercontent.com/49383323/183253332-2a471a83-6416-4a25-bf1e-f1b01abe015d.png">
<img width="541" alt="7" src="https://user-images.githubusercontent.com/49383323/183253336-6bf1bede-b688-4f79-ac74-ee6a714e8dc3.png">

- TRIMELM vs Vanilla LM (Table 2 ,3, 4)
    - 성능 향상이 존재
    - computational overhead가 없다.
- TRIMELM_long leverages long contexts
    - 결과를 보면 TRIMELM_long은 vanilla LM에 비해 큰 성능 향상을 보임
    - 단순히 vanilla LM에 continous cache를 붙이는 방식은 suboptimal한 성능
        - joint training 방식이 유효함을 암시
- TRIMELM_ext. vs KNN-LM
    - kNN-LM 모델에 비해 좋은 성능을 보임.
- Machine Translation Task
      
   <img width="540" alt="8" src="https://user-images.githubusercontent.com/49383323/183253346-287c179c-637d-4ddc-aa86-86f98d153b16.png">

    
    - 언어 모델링 뿐만아니라 번역에서도 성능 향상이 있다.

## Analysis

- Batching and memory Contruction
    - 학습과 인퍼런스 간의 gap이 없는 쪽이 좋다.
    
   <img width="554" alt="9" src="https://user-images.githubusercontent.com/49383323/183253357-aae3c629-cbb5-4588-9660-38c30dc11401.png">

    
    
- Effectiveness of using local memory
    - 로컬 메모리만을 사용하는 TRIMELM의 유효성을 검증
    - 서로 다른 세그먼트 길이 L에 따른 성능 추이
    
   <img width="555" alt="10" src="https://user-images.githubusercontent.com/49383323/183253363-751dd0db-2438-43f3-bfe7-3bc56c57a8fa.png">

    
- 기타 ablation study들은 논문 참고

# 결론

- 모델 변경 없이 long-term memory를 참조할 수 있는 TRIME 모델 제안.
- computational overhead도 없고 일관된 성능 향상을 보인다.
- 이런 방법도 self-attention의 오버헤드 때문에 긴 시퀀스를 분할하고 배치로 구성해서 참조하겠다는 건데, 빨리 성능 좋은 efficient transformer가 나와줘야…
